apps/arweave/src/ar_node_worker.erl

```diff
@@ -1046,6 +760,76 @@ apply_block(State) ->
 			apply_block(B, PrevBlocks, Timestamp, State)
 	end.
 
+may_be_rebase(State) ->
+	[{_, H}] = ets:lookup(node_state, current),
+	B = ar_block_cache:get(block_cache, H),
+	{ok, Config} = application:get_env(arweave, config),
+	case B#block.reward_addr == Config#config.mining_addr of
+		false ->
+			{noreply, State};
+		true ->
+			case ar_block_cache:get_siblings(block_cache, B) of
+				[] ->
+					{noreply, State};
+				Siblings ->
+					may_be_rebase(B, Siblings, State)
+			end
+	end.
+
+may_be_rebase(_B, _Siblings, #{ pending_rebase := true } = State) ->
+	% TODO schedule next validation step or call handle_found_solution
+	{noreply, State};
+may_be_rebase(_B, [], State) ->
+	{noreply, State};
+may_be_rebase(#block{ cumulative_diff = CDiff } = B,
+		[#block{ cumulative_diff = CDiff } = Sib | Siblings], State) ->
+	#block{ nonce_limiter_info = Info } = B,
+	#block{ nonce_limiter_info = SibInfo } = Sib,
+	StepNumber = Info#nonce_limiter_info.global_step_number,
+	SibStepNumber = SibInfo#nonce_limiter_info.global_step_number,
+	case {StepNumber > SibStepNumber, Sib#block.reward_addr == B#block.reward_addr} of
```

Comment:
The idea here is if the sibling block has a *lower* `global_step_number` we want to switch to the sibling branch and then rebuild a new block using teh current solution and apply that block onto the sibling?

How come we don't have to check cumulative diff? Are we guaranteed that 2 blocks at the same height have the same cumulative difficulty?

Reply:
No two blocks at the same time may have different cdiff e.g. during retargets.

There is a pattern matching in the argument deconstruction; perhaps, I should make it more explicit.

Reply:
Ah I see it now. Thanks!